nimsuggest/nimsuggest.nim

```diff
@@ -590,6 +594,23 @@ proc mainThread(graph: ModuleGraph) =
       conf.suggestionResultHook = proc (s: Suggest) = discard
       recompileFullProject(graph)
 
+when defined(posix):
+  proc monitorClientProcessIdThreadProc(pid: int) {.thread.} =
+    while true:
+      sleep(1000)
+      try:
+        sendSignal(Pid(pid), 0)
+      except:
+        discard kill(Pid(getCurrentProcessId()), cint(SIGTERM))
+
+when defined(windows):
+  proc monitorClientProcessIdThreadProc(pid: int) {.thread.} =
+    var process = openProcess(SYNCHRONIZE, 0, DWORD(pid))
+    if process != 0:
+      discard waitForSingleObject(process, INFINITE)
+      discard closeHandle(process)
+    quit(0)
```

Comment:
I don't understand this line.

Reply:
My intention was to handle the case, where the client process suddenly crashed, before this code gets the chance to be executed. I assume that openProcess returns an error (a NULL value) in case the process ID doesn't exist. However, I realize, I haven't tested this. So, I will check whether it's true or not.

Reply:
Ok, tested this under Windows 11. If childProcessId is nimsuggest's parent process, then openProcess will always return a handle (not NULL), even if the parent process exited, before nimsuggest could call openProcess. In this case (process has exited), the handle will be in a signalled state, so waitForSingleObject will not wait and exit immediately. In case the client process is still running, waitForSingleObject will wait for it to exit.

However, in case childProcessId is not nimsuggest's parent (e.g. it could be the process ID of the IDE, or one of the many process IDs of the IDE, like with Visual Studio Code, which spawns many processes), if the process has exited, before openProcess is called, then openProcess will return NULL and GetLastError will return ERROR_INVALID_PARAMETER. In such cases, we don't want to call waitForSingleObject on a NULL handle and we just want to exit immediately. Hence the `if process != 0:`